src/components/svg-canvas-graph.tsx

```diff
@@ -59,20 +60,22 @@ const SvgCanvas = () => {
 
         dispatch(setActive(node));
         // details panel only, remove all if this is not a multiple selection
-        if (!e.shiftKey && selected.length <= 1) dispatch(clearSelected());
-        dispatch(addSelected(node)); // details panel only
+        if (!e.shiftKey) dispatch(clearSelected());
```

Comment:
Maybe it's a bit hard to find out the relation here. Need some time to think about different situations and what the operation should be. How about writing it to

```tsx
if (!e.shiftKey) dispatch(clearSelected()); // remove all if this is not a multiple selection
else {
  if (selected.includes(node)) dispatch(removeSelected(node)); // remove current if it is already in the multiple selection
  else dispatch(addSelected(node)); // add current in the multiple selection
}
```

So basically we have 3 possible operations and none of them could overlap each other. The current code makes me wonder if `if` could be fired one by one :(

Reply:
Not quite sure what to put in the if. Any code suggestion?

Reply:
Hmm, there may be a problem with clearing the selected. We cannot move nodes that are selected if is more than one. We can delete the clear operation to avoid this.

Reply:
To be like this:

```ts
        if (!e.shiftKey && selected.length <= 1) dispatch(clearSelected()); // remove all if this is not a multiple selection
        if (e.shiftKey && selected.includes(node))
            dispatch(removeSelected(node)); // remove current if it is already in the multiple selection
        else dispatch(addSelected(node)); // add current in the multiple selection
```

Reply:
To sum up:

- `selected.includes(node) && !e.shiftKey` => do addSelected. Or we cannot deal with a single click.
- `selected.length <= 1` is needed, or we cannot move nodes together.

Reply:
Sorry for the delay. I reviewed all possible cases regarding `e.shiftKey` and `selected.includes(node)`. When classified by `e.shiftKey`, it should be non multiple selection case for `e.shiftKey === false` and multiple selection case vice versa. And for each case, we discuss the `selected.includes(node)` state. Hope the following comments in each case helps you understanding all the possible cases. Correct me if it does not behave as intended or has bugs :)

```tsx
if (!e.shiftKey) {
    // no shift key -> non multiple selection case
    if (!selected.includes(node)) {
        // set the current as the only one no matter what the previous selected were
        dispatch(setSelected([node]));
    } else {
        // no-op for clicking on the previously selected node
    }
} else {
    // shift key pressed -> multiple selection case
    if (selected.includes(node)) {
        // remove current if it is already in the multiple selection
        dispatch(removeSelected(node));
    } else {
        // add current in the multiple selection
        dispatch(addSelected(node));
    }
}
```

Reply:
PS: There might be some performance issue regarding `includes` on a large array. We might convert `selected` to `Set<...>` in the runtime slice.

Reply:
Converting `selected: string[]` to `selected: Set<string>` may be put off as too many `select.includes()` used in many files should be changed to `select.has()`. QAQ.

Reply:
That’s okay not to have `Set` in this pr. We may address this later in another issue.

Reply:
Also it looks like we can always `setSelected([node])` in non multiple selection case 🤔

Reply:
Oh, we can not `setSelected([node])` when the user clicks a previously selected one without holding shift. Previous `selected` is used to move multiple selected nodes. This should be left in the comment :(

Reply:
> Oh, we can not `setSelected([node])` when the user clicks a previously selected one without holding shift. Previous `selected` is used to move multiple selected nodes. This should be left in the comment :(
> 
> 

A, B, C are selected and I just click A again without press the shift key. We should clear selection and add A only? Is that?

Reply:
A, B, C are selected and I just click A again without pressing the shift key. -> Then we should not clear the multiple selections as users may drag before releasing the click and the selected nodes should be moved according to the drag.

So `// no-op` is correct but a better comment should be left as `// no-op as users may drag the previously selected node(s) in handlePointerMove`.

Reply:
Oh, I got it.